### DISCUSSION TITLE
how to run onedrive service at boot (no login)

### DISCUSSION THREAD

**Discussion by cholzer79** [abraunegg/onedrive#2048] [Q&A] [answered]
I am doing a clean install right now. 
On my old install I have onedrive run as a service but I am required to login for the service to start and onedrive to sync.

On my new Ubuntu VM install I'd like to have onedrive start at boot, but I don't know which service configuration achieves this.
Any help would be highly appreciated! :)

https://github.com/abraunegg/onedrive/blob/master/docs/advanced-usage.md#configuring-the-client-to-use-multiple-onedrive-accounts--configurations

**Comment by abraunegg** [comment]
@cholzer79 
Please re-read https://github.com/abraunegg/onedrive/blob/master/docs/advanced-usage.md#7-automatic-syncing-of-new-onedrive-configuration 

Tip:
* you are using ubuntu ... 
* you dont want to run the service as root

**Reply by cholzer79** [reply]
Thx, I already read that but it does not tell me (linux noob) how to get the service to run without the user having to login.
This is an ubuntu server vm, no user logs in there.

**Reply by cholzer79** [reply]
really tried following the guide in the wiki but I cannot get it to work. :(

tried to go the onedrive@.service route but I just keep getting an error that I should use `--machine=<user>.host`

could really need some help if possible. :)

goal is to have this ubuntu server VM sync 2 onedrive accounts to a mapped network share. everything works (including webhooks) I just fail at setting up these services to run at boot. :(

**Reply by abraunegg** [reply]
@cholzer79 
> could really need some help if possible. :)

8hrs ago is ~2am my time .. not everyone is based in Europe or the US :)

Things for you to test | try:
1. Ensure you have 2 x separate user accounts - login via SSH as each user
2. As each user - instead of using the @service .. which may explain your service not starting .. use `systemctl --user enable onedrive-my-new-config; systemctl --user start onedrive-my-new-config` as per https://github.com/abraunegg/onedrive/blob/master/docs/advanced-usage.md#custom-systemd-service-on-arch-ubuntu-debian-opensuse-fedora

The extreme option would be to have an entry in your rc.local file that manually starts the applications for you on reboot.

Failing this - you are using ~~Crapbuntu~~ Ubuntu - so you may need to seek systemd support from the folk that provide the distribution - as this is something I am unable to assist with further.

The last option here would be to ditch your current process, and run each 'service' as separate Docker container - as depending on how Docker is configured - those containers will startup at

**Reply by cholzer79** [reply]
Thank you for your reply!

> 2\. use `systemctl --user enable onedrive-my-new-config; systemctl --user start onedrive-my-new-config` as per https://github.com/abraunegg/onedrive/blob/master/docs/advanced-usage.md#custom-systemd-service-on-arch-ubuntu-debian-opensuse-fedora

The problem I ran into with that before is that it does not seem to use the config dir I set in the service file.

`ExecStart=/usr/bin/onedrive --confdir="/etc/onedrive/chris/" --monitor
`

If I read the log correctly, it tries to use ~/.config/onedrive/ instead.

`Jul 19 05:11:40 onedrive onedrive[905]: The database cannot be opened. Please check the permissions of ~/.config/onedrive/items.sqlite3`

I've had the same problem in Alpine Linux yesterday night when I tried to use something other than Crapuntu. 😅

**Comment by abraunegg** [comment]
@cholzer79 
> The problem I ran into with that before is that it does not seem to use the config dir I set in the service file

This is a bug that was fixed post the v2.4.19 release. See: 
https://github.com/abraunegg/onedrive/commit/e422adb4774d581d506070d9f04c947c248bee79

If you simply used the users config folder - which is the default `~/.config/onedrive` you do not get this issue. You are running foul because of using `/etc/onedrive/chris/` etc

If you are using packages, you will have to wait until v2.4.20 is released. If you build from source, or use the Docker 'edge' tag - you will not get that issue.

**Reply by cholzer79** [reply]
> This is a bug that was fixed post the v2.4.19 release

Ahhh! I am using v2.4.19-1+np1 here, I will wait for the next release then. Thank you! 

Dinner and beer coming your way! :)

btw. I also got it to work going the rc.local route (only tested with 1 account though - I am surly butchering things here XD )

**rc.local**
`sh /etc/onedrive/start-onedrive.sh`

**/etc/onedrive/start-onedrive.sh**
```
#!/bin/bash
PATH=/usr/local/sbin:/usr/sbin:/sbin:/usr/local/bin:/usr/bin:/bin
su - chris -c "/usr/bin/onedrive --confdir="/etc/onedrive/chris/" --monitor"
```

**Comment by cholzer79** [comment]
@abraunegg  sorry to bother you again, but another clean install later using `--confdir="~/.config/OneDriveChris"` I did what you suggested below

> Things for you to test | try:
> 
> 1. Ensure you have 2 x separate user accounts - login via SSH as each user
> 2. As each user - instead of using the @service .. which may explain your service not starting .. use `systemctl --user enable onedrive-my-new-config; systemctl --user start onedrive-my-new-config` as per https://github.com/abraunegg/onedrive/blob/master/docs/advanced-usage.md#custom-systemd-service-on-arch-ubuntu-debian-opensuse-fedora

**the service will not start until the user logs in**, which is exactly what I do not want (see thread title). 
I need the services to start on boot without any user logging into this headless VM.
is this possible?

**Comment by abraunegg** [comment, marked as answer]
@cholzer79 
> the service will not start until the user logs in, which is exactly what I do not want

This particular issue is not a 'onedrive' client issue .. more Linux | systemd | operating environment related ...

The answer for that lies here: https://unix.stackexchange.com/questions/636296/systemd-user-unit-not-started-after-reboot-until-user-logs-in
```
loginctl enable-linger <user>
```


I have tested this, and this and post reboot, the systemd service for the 'user' starts as normal without the 'user' having to login:
```
alex@ubuntu-headless:~$ loginctl enable-linger alex
alex@ubuntu-headless:~$ sudo reboot
```

![image](https://user-images.githubusercontent.com/4956234/179890061-815adbe6-c583-49ef-87be-7c09735facd5.png)
